rust/src/sans_io/read.rs

```diff
@@ -0,0 +1,1014 @@
+                if self.compressed.len() == 0 {
+                    self.compressed.reset();
+                }
+                match self.uncompressed.span(amount) {
+                    Span(b) => Ok(Span(b)),
+                    Remainder(_) => Ok(Remainder(std::cmp::min(
+                        clamp_to_usize(chunk_state.compressed_remaining),
+                        res.need,
+                    ))),
+                }
+            }
+        }
+    }
+
+    // Consume `amount` bytes of the uncompressed input buffer if enough is available. On failure,
+    // return the extra amount required as an error value.
+    fn consume(&mut self, amount: usize) -> McapResult<SpanOrRemainder> {
+        match self.load(amount)? {
+            Span(span) => {
+                self.uncompressed.mark_read(amount);
+                Ok(Span(span))
+            }
+            Remainder(remainder) => Ok(Remainder(remainder)),
+        }
+    }
+
+    // Return an InputBuf that requests `want` bytes from the input file. If reading
+    // from a chunk, reads into the compressed buffer, otherwise reads into the uncompressed buffer.
+    fn request(&mut self, want: usize) -> McapResult<Option<ReadAction>> {
+        return match &self.from {
+            File => Ok(Some(ReadAction::Fill(InputBuf {
+                buf: self.uncompressed.tail_with_size(want),
+                last_write: &mut self.last_write,
+            }))),
+            Chunk(state) => {
+                if state.decompressor.is_none() {
+                    Ok(Some(ReadAction::Fill(InputBuf {
+                        buf: self.uncompressed.tail_with_size(want),
+                        last_write: &mut self.last_write,
+                    })))
+                } else {
+                    Ok(Some(ReadAction::Fill(InputBuf {
+                        buf: self.compressed.tail_with_size(want),
+                        last_write: &mut self.last_write,
+                    })))
+                }
+            }
+        };
+    }
+}
+
+impl Default for LinearReader {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Encapsulates the action the user should take next when reading an MCAP file.
+///
+/// ```no_run
+/// use mcap::sans_io::read::ReadAction;
+/// use mcap::McapResult;
+/// use mcap::read::parse_record;
+/// fn use_action(file: &mut Box<dyn std::io::Read>, action: ReadAction) -> McapResult<()> {
+///   match action {
+///     ReadAction::Fill(mut into) => {
+///       let n = file.read(into.buf)?;
+///       into.set_filled(n);
+///     },
+///     ReadAction::GetRecord{ opcode, data } => {
+///       let record = mcap::parse_record(opcode, data)?;
+///       // do something with the record...
+///     }
+///   }
+///   Ok(())
+/// }
+/// ```
+pub enum ReadAction<'a> {
+    /// Fill the input buffer with raw bytes from the MCAP file. call `set_filled()` with the number
+    /// of new bytes after the read operation.
+    Fill(InputBuf<'a>),
```

Comment:
what's behind the decision to return an InputBuf with each "action"? for contrast, the typescript stream reader's `nextRecord` method (roughly equivalent to `next_action`) returns an optional record, and there's a separate `append` method for the user to provide more data. I see that it can be useful for the user to know a lower bound for how much data is needed, but what if it's more efficient for the user to provide more than that at once (e.g. a full page)?

Reply:
Some notes here:

### 1. This PR
- Caller doesn't need any buffers of their own. Reader has internal buffers U and C, which contain uncompressed and compressed data respectively.
- LinearReader alternates between yielding immutable record views into U or requesting more data into either U or C. Caller fulfils these requests directly with `read()` calls.
- if the caller wants to read pages at a time, they can allocate their own buffer (or use std::io::BufReader) to buffer reads before passing any data to the LinearReader.

### 2. Typescript-like `append()` API
- Caller feeds bulk data into the linear reader in a single call. This can mean one of:
  - The allocate their own large buffer A, fill it with MCAP data with a read() call (or it's a memory-mapped file), then calls `append(&[u8])` to copy it into the LinearReader's internal buffer B.
  - The caller calls `reserve(n) -> &mut [u8]` on the LinearReader, which extends the reader-owned buffer B and returns something for the caller to read() into.
- Reader processes through this data, yielding views into the reader-owned buffer B OR returning a signal that it needs more data.
- Since B contains a mix of compressed and uncompressed data, the reader would also need a second internal buffer C to decompress into.

My judgement is that (2) ends up with the same number of copied bytes, but requires fewer copy operations and may be more cache efficient.